### ISSUE TITLE
New cop idea: Use Comparable#clamp, Array#{min,max} instead of comparison operators

### ISSUE THREAD

**Issue by pocke** [rubocop/rubocop#11338]
## Is your feature request related to a problem? Please describe.

Today I suggested using `Comparable#clamp` instead of comparison operators in a private project, like the following.

```ruby
# bad - I feel it's not readable...
user_input > threshold ? threshold : user_input

# good - I can imagine the intention of this code easily!
user_input.clamp(..threshold)
```

I think RuboCop can aid with this problem, but there is no cop to detect this problem.

## Describe the solution you'd like

Add a cop to detect the comparison operators' usage. The cop suggests `user_input.clamp(..threshold)` instead of comparison operators.

### `Array#min`

The following code has the same meaning as the above example also.

```ruby
[user_input, threshold].min
```

Which of these is appropriate depends on the situation. If one of the values is a threshold, `clamp` clarifies the intention. But if both values are user inputs, `min` is better than `clamp`.

```ruby
# It's good
[user_input_a, user_input_b].min

# It's strange...
user_input_a.clamp(..user_input_b)
```

Therefore I think the cop should suggest both `min` and `clamp`.

### `Array#max` and `a.clamp(x..y)`

By the way, the cop can detect cases for `Array#max` and `clamp(lower_threshold..upper_threhold)` (or `clamp(lower_threshold, upper_threhold)`).

## Describe alternatives you've considered

nothing

## Additional context

There are performance differences between comparison operators, `clamp`, and `min`.

```ruby
require 'benchmark/ips'

require 'benchmark'
Benchmark.ips do |x|
  a = 42
  b = 43

  x.report('ops') { a > b ? b : a }
  x.report('min') { [a, b].min }
  x.report('clamp') { a.clamp(..b) }

  x.compare!
end
```

```console
$ ruby -v rubocop-clamp-min-performance.rb
ruby 3.2.0 (2022-12-25 revision a528908271) [arm64-darwin21]
Warming up --------------------------------------
                 ops     2.035M i/100ms
                 min     2.203M i/100ms
               clamp   879.718k i/100ms
Calculating -------------------------------------
                 ops     20.230M (± 1.4%) i/s -    101.765M in   5.031438s
                 min     21.694M (± 1.5%) i/s -    110.157M in   5.078819s
               clamp      8.694M (± 1.9%) i/s -     43.986M in   5.060963s

Comparison:
                 min: 21694180.2 i/s
                 ops: 20229721.8 i/s - 1.07x  (± 0.00) slower
               clamp:  8694430.4 i/s - 2.50x  (± 0.00) slower
```

Operators and `min` have almost the same speed because the VM executes `min` with an Array literal as an optimized instruction (`opt_newarray_min`).

```console
$ ruby --dump=insns -e '[a,b].min'
== disasm: #<ISeq:<main>@-e:1 (1,0)-(1,9)> (catch: false)
0000 putself                                                          (   1)[Li]
0001 opt_send_without_block                 <calldata!mid:a, argc:0, FCALL|VCALL|ARGS_SIMPLE>
0003 putself
0004 opt_send_without_block                 <calldata!mid:b, argc:0, FCALL|VCALL|ARGS_SIMPLE>
0006 opt_newarray_min                       2
0008 leave
```

I think this difference does not affect the entire application's performance. Probably it's a good idea that the document mention the performance.

**Comment by koic** [comment]
Is the readability of `Comparable#clamp` better in this case? So I wonder about the replacement to `Comparable#clamp`.

Background:

- RuboCop can parse Ruby 2.0 and above
- `Comparable#clamp` was introduced in Ruby 2.4
- `Comparable#clamp` introduced accepting Range as an argument in Ruby 2.7
- Beginless range was introduced in Ruby 2.7

So, in terms of readability, replacing the `>` and `<` operators with `min` and `max` seems more appropriate than `clamp`.

`Comparable#clamp` is may useful for comparisons in nested arrays:
https://andycroll.com/ruby/clamp-for-min-max-values/

What do you think?

**Comment by vlad-pisanov** [comment]
my 2¢ -- `#clamp()` seems more natural when it has two limits. A substitution strategy like this would make sense to me:

| original | replacement |
|----------|---------------|
| a > b ? a : b | [a, b].max |
| a < b ? a : b | [a, b].min |
| [[x, low].max, high].min | x.clamp(low, high) |

**Comment by pocke** [comment]
Thanks for your comments, @koic and @vlad-pisanov!

Personally I like `clamp` to round a value with a max or min limit, even if it takes only either max or min. But I can easily imagine your opinion, `min` or `max` are preferred for this purpose.


I prefer the following replacements.

| original | replacement |
|----------|---------------|
| `a > b ? a : b` | `[a, b].max` or `a.clamp(b..)` (depend on context) |
| `a < b ? a : b` | `[a, b].min` or `a.clamp(..b)` (depend on context) |
| `[[x, low].max, high].min` | `x.clamp(low, high)` |


I guess we have three options for `a > b ? a : b` and `a < b ? a : b`.

* Suggest `Array#max` or `Array#min` always
  * @koic and @vlad-pisanov perfer
* Suggest (`Array#max` or `Array#min`) and `Comparable#clamp`
  * @pocke prefer
  * Because I'd like to use both depending on the context
* Suggest `Comparable#clamp` always
  * I'm not sure anyone prefers it...

-----

By the way, this cop should be marked as `Safe: false`. Because even if a value has `<` method, it's not necessarily Comparable.

**Comment by koic** [comment]
> Because I'd like to use both depending on the context

Hm, are there any use cases where you prefer `a.clamp(b..)`  over `[a, b].max`?

**Comment by koic** [comment]
There is still consider about `Comparable#clamp`. On the other hand, it seems common to prefer `[a, b].max` to `a > b ? a : b`.

First, I opened PR #11341 to create a cop that enforces `[a, b].max`. Because I think it's better to have a separate cop for `Comparable#clamp`. This is because the different Ruby versions that can be targeted.

**Comment by pocke** [comment]
> > Because I'd like to use both depending on the context
> 
> Hm, are there any use cases where you prefer `a.clamp(b..)` over `[a, b].max`?

Ah, sorry, I forgot to describe the context. I think `clamp` is more appropriate if one of the values is a limit or threshold.

For example: Limit per page on pagination

```ruby
MAX_ITEMS_PER_PAGE = 100

def index
  raw_per_page = params[:per_page].to_i

  # With `clamp`
  per_page = raw_per_page.clamp(..MAX_ITEMS_PER_PAGE)

  # With `min`
  per_page = [raw_per_page, MAX_ITEMS_PER_PAGE].min

  @items = Item.all.per_page(per_page)
end
```


`min` means "Pick out minimum count from `raw_per_page` and `MAX_ITEMS_PER_PAGE`".
`clamp` means "Limit `raw_per_page` to `MAX_ITEMS_PER_PAGE`".

In this case, I prefer `clamp` to `min` because I feel `clamp` is more intentional. `min` just means minimum value, but `clamp` expresses that the argument is a limit value.

**Comment by pocke** [comment]
> Because I think it's better to have a separate cop for Comparable#clamp. This is because the different Ruby versions that can be targeted.


Well, if `a.clamp(..b)` case is also acceptable, two cops warn the same code with different messages, right?
For example, for `per_page = per_page > MAX_ITEMS_PER_PAGE ? MAX_ITEMS_PER_PAGE : per_page`, `Style/MinMaxComparison` cop suggests using `min`, but `Style/ClampComparison` cop suggests using `clamp`.

It may be confusing.

**Comment by koic** [comment]
Thanks for explaining your use case.

I wonder if it may be a bit difficult for user to choose whether to use `{min, max}` or `clamp`. So, I think it would be clear to user to provide autocorrection for `{min, max}`. `{min, max}` are traditional simple methods.
And I'm also concerned about the slowness of `clamp`. Maybe only documented for `clamp` against `# good` case and text in #11341.

So the next steps are:

- Add documentation about `clamp` to #11343
- Provides a new cop proposing `x.clamp(low, high)` instead of `[[x, low].max, high].min`. It's a different police than #11341

> It may be confusing.

This probably won't cause that confusion. OTOH, the only my concern is that `{min, max}` is chosen for use cases where `clamp` is preferable, but user can refactor it to `clamp` userself.

What do you think of this design, which is different from the original intention of the issue?

**Comment by pocke** [comment]
> So the next steps are:


It looks good, thanks for your suggestion!

> OTOH, the only my concern is that {min, max} is chosen for use cases where clamp is preferable, but user can refactor it to clamp userself.

I think so too. The user maybe doesn't notice the existence of `clamp`. I think your suggestion is a good compromise 👍 


---


> And I'm also concerned about the slowness of clamp.


`min` and `max` are also slower than the operators on old Rubies.

```console
$ docker run --rm -it rubylang/all-ruby env ALL_RUBY_BINS='ruby-2.0.0-p648 ruby-2.1.10 ruby-2.2.10 ruby-2.3.8 ruby-2.4.10 ruby-2.5.9 ruby-2.6.10 ruby-2.7.7 ruby-3.0.5 ruby-3.1.3 ruby-3.2.0' ./all-ruby -rbenchmark -we 'a=42;b=43;p Benchmark.realtime{100000.times{[a,b].min}}; p Benchmark.realtime{100000.times{a > b ? b : a}}'
ruby-2.0.0-p648 0.195302542
                0.035441417
ruby-2.1.10     0.175730625
                0.035914958
ruby-2.2.10     0.194358040811494
                0.04013233305886388
ruby-2.3.8      0.18656654190272093
                0.035134333884343505
ruby-2.4.10     0.03193425014615059
                0.03173887496814132
ruby-2.5.9      0.03606400010176003
                0.03370579099282622
ruby-2.6.10     0.03306250018067658
                0.028806084068492055
ruby-2.7.7      0.030842124950140715
                0.02505087503232062
ruby-3.0.5      0.03810958401300013
                0.02866741712205112
ruby-3.1.3      0.04034850001335144
                0.03262170893140137
ruby-3.2.0      0.028015041025355458
                0.027625084156170487
```

`min` is slower than the operators on Ruby 2.3 or older. Should we treat this difference on the cop's implementation?
I guess it does not affect the performance of an entire application in most cases. It's enough that the documentation mentions the performance.

**Comment by koic** [comment]
> min and max are also slower than the operators on old Rubies.

Hm, I think it depends on the operands. For example, `a > b ? a : b` evaluates to either `a` or `b` twice. So if `a` or `b` is a slow method call, I think `max` is better. In other words, the results may change depending on the benchmark data or usage context, right?

**Comment by pocke** [comment]
> > min and max are also slower than the operators on old Rubies.
> 
> Hm, I think it depends on the operands. For example, `a > b ? a : b` evaluates to either `a` or `b` twice. So if `a` or `b` is a slow method call, I think `max` is better. In other words, the results may change depending on the benchmark data or usage context, right?

True. The summary is the following:

* Replacing `min` has a performance risk until Ruby 2.4 because it doesn't have the optimized instruction.
  * OTOH, it may improve the performance as you said.
* Replacing `min` has no performance risk since Ruby 2.4 because it has the optimized instruction.
  * It also may improve the performance as you said.

I'm not sure the document should describe the performance. It will be helpful to decide to follow the cop. But it may too detailed.

**Comment by koic** [comment]
JFYI, this issue is solved by #11341, #11364, and #11365.
